### PR TITLE
Add new badges design with UI editor

### PR DESCRIPTION
**Pull request by piitaya**
## Breaking change

The current badges will use the new design.

### Hui Badge

Badge are now loaded using `hui-badge` component instead of having the view responsible of loading these element

### New badge 🎉

CleanShot 2024-07-17 at 10 25 45

A new type of badge has been added : the `entity` badge. The design is more compact than the previous one and better match the tile card design.

3 display types are available : 
- `minimal` : icon only
- `standard` : icon and state, this is the default
- `complete` : icon, name and state

Customization are similar to tile card options :  name, icon, color, state content and entity picture instead of icon.

### Visibility options

Like cards and sections, badges support all the available visibility conditions : https://www.home-assistant.io/dashboards/conditional/#conditions-options

### Custom badges

Custom badges was already supported but not documented. It's now supported by the card editor. It's loaded the same way as custom cards. A pull request in developer blog and documentation will follow.

### Edit mode

The badge can be added using the `+` on the top of the view. Masonry view and section view are the only ones that supports badges. Badges can also be deleted and moved using drag and drop.

By clicking a badge, you can edit all the configuration directly from the UI the same way as card.

CleanShot 2024-07-17 at 11 24 32

### Future Improvement to do (not in this PR) : 
- Add copy/cut badges
- Better badge error 
- Use badge picker when adding a badge

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33866

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

## Summary by CodeRabbit

- **New Features**
  - Added support for customizable and interactive badges in Lovelace views.
  - Introduced new `HuiBadge` and `HuiEntityBadge` elements for enhanced badge management.
  - Added `HuiViewBadges` component for flexible badge display and editing.
  - Implemented `hui-badge-edit-mode` for easier badge configuration and manipulation.

- **Improvements**
  - Enhanced badge visibility and functionality with new properties like `visibility` and `preview`.
  - Updated UI elements to better handle state representation and interactions within Lovelace.
  - Introduced new interface `CustomBadgeEntry` to streamline custom badge management.

- **Bug Fixes**
  - Fixed issues related to badge display and interaction, ensuring consistent behavior across different Lovelace views.

### CODE REVIEW

**Comment by surfingbytes** [comment]
This new design is great, thank you 👍 Much more compact, so I'm sure users will like it. However, it doesn't give the "warning" vibe as the old one, so could we please have them side by side (maybe with the old one changed to regular card, so it can be used anywhere). Here is an example how the current one can be used as "warning" to water plant and show running timer, which the new design really is not the same:
image

**Comment by piitaya** [comment]
Thank you for the feedback.

The old badges will always work as before. Badges are just a smaller card. You can mix them if you want but it can not look very good as the height is not the same for these different kind of badge. 

Also, more type of badge can be added in the future (alerting ones for example if it make sense). Even custom badges can be added.

**Comment by Gloomyeye** [comment]
The new badges look good. I would prefer if you could add an optional name/title to the badges. If you add two light badges, then you can only distingush the badges with the state or the icon. A badge name/title (maybe in a secondary line) would give more insights.

**Comment by piitaya** [comment]
@coderabbitai pause

File: src/data/lovelace/config/view.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export interface LovelaceBaseViewConfig {
 
 export interface LovelaceViewConfig extends LovelaceBaseViewConfig {
   type?: string;
-  badges?: Array<string | LovelaceBadgeConfig>;
+  badges?: LovelaceBadgeConfig[];
```
Comment (bramkragten) [inline src/data/lovelace/config/view.ts:30]: we should still support just strings, no?

File: src/panels/lovelace/badges/types.ts [hunk 2]
```diff
@@ -25,3 +26,15 @@ export interface StateLabelBadgeConfig extends LovelaceBadgeConfig {
   hold_action?: ActionConfig;
   double_tap_action?: ActionConfig;
 }
+
+export interface EntityBadgeConfig extends LovelaceBadgeConfig {
+  type: "entity";
+  entity?: string;
+  name?: string;
+  icon?: string;
```
Comment (bramkragten) [inline src/panels/lovelace/badges/types.ts:34]: The `StateLabelBadgeConfig` supported `image`, should we keep that around in the new one?

Maybe also migrate the option `show_name` to a `display_type`?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 3]
```diff
@@ -431,7 +431,7 @@ export class HuiDialogEditCard
     await this._params!.saveConfig(
       "cardConfig" in this._params!
         ? addCard(this._params!.lovelaceConfig, path, this._cardConfig!)
-        : replaceCard(
+        : replaceBadge(
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts:434]: replaceBadge? 🤔

File: src/panels/lovelace/editor/config-elements/hui-entity-badge-editor.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import { EntityBadgeConfig } from "../../badges/types";
+import type { LovelaceBadgeEditor } from "../../types";
+import "../hui-sub-element-editor";
+import { actionConfigStruct } from "../structs/action-struct";
+import { baseLovelaceCardConfig } from "../structs/base-card-struct";
+import { configElementStyle } from "./config-elements-style";
+import "./hui-card-features-editor";
+
+const badgeConfigStruct = assign(
+  baseLovelaceCardConfig,
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-entity-badge-editor.ts:37]:
```suggestion
  baseLovelaceBadgeConfig,
```

**Review by bramkragten — CHANGES_REQUESTED** [review]

File: src/data/lovelace/config/section.ts [hunk 5]
```diff
@@ -10,6 +11,7 @@ export interface LovelaceBaseSectionConfig {
 export interface LovelaceSectionConfig extends LovelaceBaseSectionConfig {
   type?: string;
   cards?: LovelaceCardConfig[];
+  badges?: LovelaceBadgeConfig[]; // Not supported yet
```
[the remark this answers is not in the document]
Comment (piitaya) [reply src/data/lovelace/config/section.ts]: I removed it

File: src/data/lovelace/config/view.ts (hunk 1 above)
Comment (piitaya) [reply src/data/lovelace/config/view.ts]: Yes for backward compatibility, I reverted that change.

File: src/panels/lovelace/badges/types.ts (hunk 2 above)
Comment (piitaya) [reply src/panels/lovelace/badges/types.ts:34]: I will make another PR to properly migrate the state label badge

**Review by bramkragten — APPROVED** [review]

**Pull request merged**